src/test/java/org/elasticsearch/example/realm/CustomCachingRealmTests.java

```diff
@@ -52,28 +55,34 @@ public void testAuthenticateWithCachedValue() {
                 new Environment(globalSettings), new ThreadContext(globalSettings)));
 
         // authenticate john
-        UsernamePasswordToken token = new UsernamePasswordToken("john", new SecuredString(new char[] { 'd', 'o', 'e'}));
-        final User user = realm.authenticate(token);
-        assertThat(user, notNullValue());
-        assertThat(user.roles(), arrayContaining("user"));
-        assertThat(user.principal(), equalTo("john"));
+        UsernamePasswordToken token = new UsernamePasswordToken("john", new SecureString(new char[] { 'd', 'o', 'e'}));
+        realm.authenticate(token, ActionListener.wrap(user -> {
+            assertThat(user, notNullValue());
+            assertThat(user.roles(), arrayContaining("user"));
+            assertThat(user.principal(), equalTo("john"));
+            CustomCachingRealmTests.this.first = user;
```

Comment:
We generally use an AtomicReference (local to the test method) for this sort of testing, rather than a field.